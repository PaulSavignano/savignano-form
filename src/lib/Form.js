import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import FormContext from './FormContext'
import setIn from './utils/setIn'
import getIn from './utils/getIn'
import isValue from './utils/isValue'
import handleSubmitEvent from './utils/handleSubmitEvent'
import handleInputEvent from './utils/handleInputEvent'

class Form extends PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      errors: {},
      initialValues: {},
      isSubmitSuccess: false,
      isSubmitting: false,
      isTouched: false,
      submitError: '',
      touched: {},
      values: {},
    }
    this.fields = {}
    this.mounted = false
  }

  componentDidMount() {
    this.mounted = true
  }

  componentWillUnmount() {
    this.mounted = false
  }

  getOnProps = props => ({
    ...this.state,
    ...props,
    getIn,
    onBlur: this.handleOnBlur,
    onChange: this.handleOnChange,
    onError: this.handleOnError,
  })

  handleBlur = ({ name, value }) => {
    const { onBlur } = this.fields[name]
    this.setState(state => ({
      isTouched: true,
      touched: setIn(state.touched, name, true),
    }))
    if (onBlur) {
      return onBlur(this.getOnProps({ name, value }))
    }
    return undefined
  }

  handleBlurEvent = e => {
    const { name, value } = handleInputEvent(e)
    this.handleBlur({ name, value })
  }

  handleChange = ({ name, value }) => {
    const { onChange, onParse, onValidate } = this.fields[name]
    const parsedValue = this.handleOnParse({ name, onParse, value })
    const error = this.handleOnValidate({ name, onValidate, value })
    this.setState((state) => {
      const errors = this.props.onValidate ? this.props.onValidate(state) : state.errors
      return {
        errors: setIn(errors, name, error),
        isTouched: true,
        submitError: '',
        values: setIn(state.values, name, parsedValue),
      }
    })
    if (onChange) {
      return onChange(this.getOnProps({ name, value }))
    }
    return undefined
  }

  handleChangeEvent = (e) => {
    const { name, value, checked } = handleInputEvent(e)
    const { type } = this.fields[name]
    const isCheckbox = /checkbox/.test(type)
    return this.handleChange({ name, value: isCheckbox ? checked : value })
  }

  handleOnBlur = ({ name }) => this.setState(state => ({
    isTouched: true,
    touched: setIn(state.touched, name, true),
  }))

  handleOnChange = ({ name, value }) => this.setState(state => ({
    values: setIn(state.values, name, value),
  }))

  handleOnError = ({ name, error }) => this.setState(state => ({
    errors: setIn(state.errors, name, error),
  }))

  handleOnParse = ({ name, onParse, value }) => {
    if (onParse) {
      return onParse(this.getOnProps({ name, value }))
    }
    return value
  }

  handleOnValidate = ({ name, onValidate, value }) => {
    if (!onValidate) return undefined
    if (Array.isArray(onValidate)) {
      return onValidate.reduce((a, v) => {
        let error = a
        error = v(this.getOnProps({ name, value }))
        return error
      }, {})
    }
    return onValidate(this.getOnProps({ name, value }))
  }

  handleRegisterField = ({ name, ...rest }) => {
    if (this.fields[name]) return undefined
    this.fields[name] = rest
    const initialValue = getIn(this.props.initialValues, name) || getIn(this.props.defaultValues, name)
    if (isValue(initialValue)) {
      return this.setState(state => ({
        initialValues: setIn(state.values, name, initialValue),
        values: setIn(state.values, name, initialValue),
      }))
    }
    return undefined
  }

  handleUnregisterField = ({ name }) => {
    this.setState(state => ({
      errors: setIn(state.errors, name, undefined),
      touched: setIn(state.touched, name, undefined),
      values: setIn(state.values, name, undefined),
    }))
    delete this.fields[name]
  }

  handleReset = (names) => {
    if (names && names.length) {
      const nextState = Object.keys(this.fields).reduce((a, name) => {
        const state = a
        if (names.includes(name)) {
          const value = this.handleResetValue(name)
          state.errors = setIn(state.errors, name, undefined)
          state.initialValues = setIn(
            state.initialValues,
            name,
            value,
          )
          state.touched = setIn(state.touched, name, undefined)
          state.values = setIn(state.values, name, value)
          return state
        }
        state.errors = setIn(state.errors, name, getIn(this.state.errors, name))
        state.initialValues = setIn(state.initialValues, name, this.handleResetValue(name))
        state.touched = setIn(state.touched, name, getIn(this.state.touched, name))
        state.values = setIn(state.values, name, getIn(this.state.values, name))
        return state
      }, {})
      return this.setState({
        ...nextState,
        isTouched: Boolean(Object.keys(nextState.touched).length),
      })
    }
    const nextState = Object.keys(this.fields).reduce((a, name) => {
      const state = a
      const initialValue = this.handleResetValue(name)
      const value = this.handleResetValue(name)
      state.initialValues = setIn(
        state.initialValues,
        name,
        initialValue,
      )
      state.touched = setIn(
        state.touched,
        name,
        undefined,
      )
      state.values = setIn(state.values, name, value)
      return state
    }, {})
    return this.setState({
      ...nextState,
      errors: {},
      isSubmitSuccess: false,
      isSubmitting: false,
      isTouched: Boolean(Object.keys(nextState.touched).length),
      submitError: '',
    })
  }

  handleResetValue = (name) => {
    const initialValue = getIn(this.props.initialValues, name)
    const defaultValue = getIn(this.props.defaultValues, name)
    if (isValue(initialValue)) return initialValue
    if (isValue(defaultValue)) return defaultValue
    return undefined
  }

  handleState = state => {
    return this.setState(state)
  }

  handleSubmitValidations = () => {
    const validationState = Object.keys(this.fields).reduce((a, name) => {
      const { onValidate } = this.fields[name]
      const error = onValidate
        && this.handleOnValidate({
          name,
          value: getIn(this.state.values, name),
          onValidate,
        })
      a.touched = setIn(a.touched, name, true)
      if (error) {
        a.errors = setIn(a.errors, name, error)
      }
      return a
    }, {})
    return validationState
  }

  handleSubmit = (e) => {
    const { onSubmit: onSubmitProp } = this.props
    const fn = onSubmitProp || e
    console.log('handleSubmit fn ', fn)
    handleSubmitEvent(e)
    this.setState({ isSubmitting: true })
    const validationState = this.handleSubmitValidations()
    if (validationState.errors && Object.keys(validationState.errors).length) {
      return this.setState({ ...validationState, isSubmitting: false })
    }
    return fn(this.state.values)
      .then((res) => {
        if (this.mounted) {
          this.setState({ isSubmitSuccess: true })
          this.handleReset()
        }
        return Promise.resolve(res)
      })
      .catch((error) => {
        if (this.mounted) this.setState({ isSubmitting: false, submitError: error })
      })
  }

  render() {
    const {
      children,
      component: Comp,
      onSubmit,
      initialValues,
      defaultValues,
      onValidate,
      ...rest
    } = this.props
    const ctx = {
      ...this.state,
      isErrors: Boolean(Object.keys(this.state.errors).length),
      onBlur: this.handleBlurEvent,
      onChange: this.handleChangeEvent,
      onRegisterField: this.handleRegisterField,
      onReset: this.handleReset,
      onState: this.handleState,
      onSubmit: this.handleSubmit,
      onUnregisterField: this.handleUnregisterField,
      formProps: this.props,
    }
    return (
      <FormContext.Provider value={ctx}>
        <Comp
          {...rest}
          {...onSubmit && { onSubmit: this.handleSubmit }}
        >
          {children}
        </Comp>
      </FormContext.Provider>
    )
  }
}

Form.defaultProps = {
  className: undefined,
  component: 'form',
  defaultValues: {},
  initialValues: {},
  onSubmit: undefined,
  onValidate: undefined,
  style: undefined,
}

Form.propTypes = {
  children: PropTypes.oneOfType([PropTypes.func, PropTypes.node]).isRequired,
  className: PropTypes.string,
  component: PropTypes.oneOfType([PropTypes.func, PropTypes.node, PropTypes.string]),
  defaultValues: PropTypes.shape(Object),
  initialValues: PropTypes.shape(Object),
  onSubmit: PropTypes.func,
  onValidate: PropTypes.func,
  style: PropTypes.objectOf(PropTypes.object),
}

export default Form
