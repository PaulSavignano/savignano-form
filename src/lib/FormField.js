import React, { useMemo } from 'react'
import PropTypes from 'prop-types'

import useFormField from './useFormField'

function FormField({
  name,
  id,
  onBlur,
  onChange,
  onFormat,
  onParse,
  onValidate,
  type,
  value,
  isPersistOnUnmount,
  component: Component,
  ...rest
}) {
  const fieldProps = useFormField({
    name,
    id,
    onBlur,
    onChange,
    onFormat,
    onParse,
    onValidate,
    isPersistOnUnmount,
    type,
    value,
  })
  return useMemo(() => {
    const { isTouched, ...restFieldProps } = fieldProps
    const isTouchedProp = typeof Component === 'string' ? {} : { isTouched }
    return (
      <Component {...rest} {...restFieldProps} {...isTouchedProp} />
    )
  }, [
    value,
    fieldProps.value,
    fieldProps.checked,
    fieldProps.error,
    fieldProps.isTouched,
  ])
}

FormField.defaultProps = {
  id: undefined,
  onBlur: undefined,
  onChange: undefined,
  onFormat: undefined,
  onParse: undefined,
  onValidate: undefined,
  type: undefined,
  value: undefined,
  isPersistOnUnmount: false
}

FormField.propTypes = {
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.node,
    PropTypes.string
  ]).isRequired,
  id: PropTypes.string,
  name: PropTypes.string.isRequired,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFormat: PropTypes.func,
  onParse: PropTypes.func,
  onValidate: PropTypes.oneOfType([PropTypes.func, PropTypes.array]),
  type: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.bool, PropTypes.number, PropTypes.string])
}

export default FormField
