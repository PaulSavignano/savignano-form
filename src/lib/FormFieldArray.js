import React, { useContext, useState } from 'react'
import PropTypes from 'prop-types'

import FormContext from './FormContext'
import getIn from './utils/getIn'
import setIn from './utils/setIn'

export function getValueArray(value) {
  if (value && value.length) {
    return value.map((v, i) => i)
  }
  return []
}

export function getInitialValueArray({ name, values, initialValues, defaultValues }) {
  const value = getIn(values, name)
  if (value) return getValueArray(value)
  const initialValue = getIn(initialValues, name)
  if (initialValue) return getValueArray(initialValue)
  const defaultValue = getIn(defaultValues, name)
  if (defaultValue) return getValueArray(defaultValue)
  return []
}

export function getFiltered({ index, value }) {
  if (value && value.length) {
    return value.filter((v, i) => i !== index)
  }
  return undefined
}

function FormFieldArray({ name, component: Component, ...rest }) {
  const { onState, values, formProps } = useContext(FormContext)
  const value = getIn(values, name) || []
  const [valueArray, setValueArray] = useState(getInitialValueArray({ name, values, initialValues: formProps.initialValues, defaultValues: formProps.defaultValues }))
  return (
    <Component
      {...rest}
      name={name}
      onAdd={() => {
        return setValueArray([...valueArray, valueArray.length])
      }}
      onDelete={(index) => {
        const filtered = getFiltered({ index, value })
        onState({ values: setIn(values, name, filtered) })
        return setValueArray(valueArray.filter((v, i) => i !== index))
      }}
      onChange={(changedValue) => {
        const nextValues = setIn(values, name, changedValue)
        const nextValueArray = getValueArray(changedValue)
        onState({ values: nextValues })
        return setValueArray(nextValueArray)
      }}
      value={valueArray}
    />
  )
}

FormFieldArray.propTypes = {
  name: PropTypes.string.isRequired,
}

export default FormFieldArray
