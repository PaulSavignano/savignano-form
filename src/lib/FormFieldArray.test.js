import React from 'react'
import {
  cleanup,
  fireEvent,
  render,
} from '@testing-library/react'

import Form from './Form'
import FormField from './FormField'
import FormFieldArray, { getInitialValueArray, getFiltered, getValueArray } from './FormFieldArray'

const initialValues = {
  emails: [{ email: 'test1@test.com' }]
}

const TestComponent = ({
  name,
  onAdd,
  onChange,
  onDelete,
  value,
}) => {
  console.log('TestComponent', value)
  return (
    <div>
      <button onClick={() => onAdd()} type="button">
        Add
      </button>
      <button
        onClick={() => onChange([{ email: 'test1@test.com' }, { email: 'test2@test.com' }])}
        type="button"
      >
        Change
      </button>
      <ul>
        {value.map((v, i) => (
          <li key={i}>
            <FormField
              name={`${name}.${i}.email`}
              component="input"
              label={`${name} ${i}`}
              id={`${name} ${i}`}
            />
            <button data-testid={`delete-${i}`} onClick={() => onDelete(i)} type="button">
              Delete
            </button>
          </li>
        ))}
      </ul>
    </div>
  )
}


afterEach(cleanup)

describe('getValueArray', () => {
  it('should return index array from object array', () => {
    const valueArray = getValueArray([{ email: 'test1@test.com' }, { email: 'test2@test.com' }])
    expect(valueArray).toEqual([0, 1])
  })
  it('should return empty array if value is undefined', () => {
    const valueArray = getValueArray(undefined)
    expect(valueArray).toEqual([])
  })
})

describe('getInitialValueArray', () => {
  it('should return array of indexes', () => {
    const values = { emails: [{ email: 'test1@test.com' }, { email: 'test2@test.com' }] }
    const initialValueArray = getInitialValueArray({ name: 'emails', values })
    expect(initialValueArray).toEqual([0, 1])
  })
  it('should return empty array if value is undefined', () => {
    const values = {}
    const initialValueArray = getInitialValueArray({ name: 'emails', values })
    expect(initialValueArray).toEqual([])
  })
  it('should return defaultValue indexes if no value or initialValue', () => {
    const defaultValues = { emails: [{ email: 'test1@test.com' }, { email: 'test2@test.com' }] }
    const initialValueArray = getInitialValueArray({ name: 'emails', defaultValues })
    expect(initialValueArray).toEqual([0, 1])
  })
})

describe('getFiltered', () => {
  it('should filter array by index', () => {
    const props = {
      index: 0,
      value: [{ email: 'test1@test.com' }, { email: 'test2@test.com' }]
    }
    const filtered = getFiltered(props)
    expect(filtered).toEqual([props.value[1]])
  })
  it('should return undefined if no value', () => {
    const props = {
      index: 0,
      value: undefined
    }
    const filtered = getFiltered(props)
    expect(filtered).toEqual(undefined)
  })
})

describe('FormFieldArray', () => {
  it('should match snapshot', async () => {
    const { asFragment } = render(
      <Form initialValues={initialValues}>
        <FormFieldArray name="emails" component={TestComponent} />
      </Form>
    )
    expect(asFragment()).toMatchSnapshot()
  })


  it('should add to array when calling onAdd', () => {
    const { getByText, queryAllByText } = render(
      <Form initialValues={initialValues}>
        <FormFieldArray name="emails" component={TestComponent} />
      </Form>
    )
    fireEvent.click(getByText("Add"))
    const deleteButtons = queryAllByText('Delete')
    expect(deleteButtons).toHaveLength(2)
  })

  it('should delete from array when calling onDelete', () => {
    const { getByTestId, queryAllByText } = render(
      <Form initialValues={initialValues}>
        <FormFieldArray name="emails" component={TestComponent} />
      </Form>
    )
    fireEvent.click(getByTestId('delete-0'))
    const deleteButtons = queryAllByText('Delete')
    expect(deleteButtons).toHaveLength(0)
  })

  it('should change root array to new objects', () => {
    const { getByText, queryAllByText } = render(
      <Form initialValues={initialValues}>
        <FormFieldArray name="emails" component={TestComponent} />
      </Form>
    )

    fireEvent.click(getByText("Change"))
    const deleteButtons = queryAllByText('Delete')
    expect(deleteButtons).toHaveLength(2)
  })


})
