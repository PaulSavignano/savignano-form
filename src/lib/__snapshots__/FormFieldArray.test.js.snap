// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFieldArray should match snapshot 1`] = `
<DocumentFragment>
  <form>
    <div>
      <button
        type="button"
      >
        Add
      </button>
      <button
        type="button"
      >
        Change
      </button>
      <ul>
        <li>
          <input
            aria-label="emails 0"
            name="emails.0.email"
            type="text"
            value="test1@test.com"
          />
          <button
            data-testid="delete-0"
            type="button"
          >
            Delete
          </button>
        </li>
      </ul>
    </div>
  </form>
</DocumentFragment>
`;
